plugins/askrene/refine.c

```diff
@@ -0,0 +1,379 @@
+ * increased (which sets reservations) */
+static struct amount_msat flow_remaining_capacity(struct route_query *rq,
+						  const struct flow *flow)
+{
+	struct amount_msat max_msat = AMOUNT_MSAT(-1ULL);
+	for (int i = tal_count(flow->path) - 1; i >= 0; i--) {
+		const struct half_chan *h = flow_edge(flow, i);
+		struct amount_msat min, max;
+
+		/* We can pass constraints due to addition of fees! */
+		get_constraints(rq, flow->path[i], flow->dirs[i], &min, &max);
+		max = amount_msat_min(max, amount_msat(fp16_to_u64(h->htlc_max)));
+
+		max_msat = amount_msat_min(max_msat, max);
+		if (!amount_msat_add_fee(&max_msat, h->base_fee, h->proportional_fee))
+			max_msat = AMOUNT_MSAT(-1ULL);
+	}
+
+	/* Calculate deliverable max */
+	for (size_t i = 0; i < tal_count(flow->path); i++) {
+		const struct half_chan *h = flow_edge(flow, i);
+		max_msat = amount_msat_sub_fee(max_msat,
+					       h->base_fee, h->proportional_fee);
+	}
+	return max_msat;
+}
+
+/* What's the "best" flow to add to? */
+static struct flow *pick_most_likely_flow(struct route_query *rq,
+					  struct flow **flows,
+					  struct amount_msat additional)
+{
+	double best_prob = 0;
+	struct flow *best_flow = NULL;
+
+	for (size_t i = 0; i < tal_count(flows); i++) {
+		struct amount_msat cap;
+		double prob = flow_probability(flows[i], rq);
+		if (prob < best_prob)
+			continue;
+		cap = flow_remaining_capacity(rq, flows[i]);
+		if (amount_msat_less(cap, additional))
+			continue;
+		best_prob = prob;
+		best_flow = flows[i];
+		plugin_log(rq->plugin, LOG_DBG, "Best flow is #%zu!", i);
+	}
+
+	return best_flow;
+}
+
+const char *
+refine_with_fees_and_limits(const tal_t *ctx,
+			    struct route_query *rq,
+			    struct amount_msat deliver,
+			    struct flow ***flows)
+{
+	struct reservations *reservations = new_reservations(NULL, rq);
+	struct amount_msat more_to_deliver;
+	const char *flow_constraint_error = NULL;
+	const char *ret;
+
+	for (size_t i = 0; i < tal_count(*flows);) {
+		struct flow *flow = (*flows)[i];
+
+		plugin_log(rq->plugin, LOG_DBG, "Constraining flow %zu: %s",
+			   i, fmt_amount_msat(tmpctx, flow->delivers));
+		for (size_t j = 0; j < tal_count(flow->path); j++) {
+			struct amount_msat min, max;
+			get_constraints(rq, flow->path[j], flow->dirs[j], &min, &max);
+			plugin_log(rq->plugin, LOG_DBG, "->%s(max %s)",
+				   fmt_flows_step_scid(tmpctx, rq, flow, j),
+				   fmt_amount_msat(tmpctx, max));
+		}
+
+		flow_constraint_error = constrain_flow(tmpctx, rq, flow, reservations);
+		if (!flow_constraint_error) {
+			i++;
+			continue;
+		}
+
+		plugin_log(rq->plugin, LOG_DBG, "Flow was too constrained: %s",
+			   flow_constraint_error);
+		/* This flow was reduced to 0 / impossible, remove */
+		tal_arr_remove(flows, i);
+	}
+
+	/* Due to granularity of MCF, we can deliver slightly more than expected:
+	 * trim one in that case. */
+	if (!amount_msat_sub(&more_to_deliver, deliver,
```

Comment:
Here we remove some excess due to the granularity of MCF, this means we might reduce the amount we send with a flow, that might trigger an `htlc_min` constraint.

Reply:
True!  I'll add *ANOTHER* check for that...